File: kotlin/src/main/kotlin/jomiel/examples/Jomiel.kt
/*
 * -*- coding: utf-8 -*-
 *
 * jomiel-examples
 *
 * Copyright
 *  2021 Toni Gündoğdu
 *
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package jomiel.examples

import jomiel.protobuf.v1beta1.*
import jomiel.protobuf.v1beta1.Inquiry.Inquiry.Media
import org.tinylog.kotlin.Logger.error
import org.tinylog.kotlin.Logger.info
import org.zeromq.SocketType.REQ
import org.zeromq.ZContext
import org.zeromq.ZMQ.Poller.POLLIN
import pbandk.ExperimentalProtoJson
import pbandk.Message
import pbandk.decodeFromByteArray
import pbandk.encodeToByteArray
import pbandk.json.JsonConfig.Companion.DEFAULT
import pbandk.json.encodeToJsonString
import kotlin.system.exitProcess

class Jomiel(private val opts: Runner) {
    private val ctx = ZContext()
    private val sck = ctx.createSocket(REQ)
    private val poller = ctx.createPoller(1)

    @OptIn(ExperimentalProtoJson::class)
    private val jsonConfig =
        DEFAULT.copy(compactOutput = opts.compactJson)

    init {
        sck.linger = 0
        poller.register(sck, POLLIN)
    }

    fun run() {
        when {
            !opts.uri.isNullOrEmpty() -> {
                connect()
                opts.uri.forEach { inquire(it) }
            }
            else -> {
                error("error: input URI not given")
                exitProcess(1)
            }
        }
    }

    private fun connect() {
        val re = opts.routerEndpoint
        val to = opts.connectTimeout
        printStatus("<connect> $re (timeout=$to)")
        sck.connect(re)
    }

    private fun inquire(uri: String) {
        sendInquiry(uri)
        receiveResponse()
    }

    private fun sendInquiry(uri: String) {
        val msg = Inquiry(
            inquiry = Media(
                media = MediaInquiry(inputUri = uri)
            )
        )
        if (!opts.beTerse) printMessage("<send>", msg)
        val bytes = msg.encodeToByteArray()
        sck.send(bytes)
    }

    private fun receiveResponse() {
        poller.poll(opts.connectTimeout * 1_000)
        when {
            poller.pollin(0) -> {
                val bytes = sck.recv()
                val msg = Response.decodeFromByteArray(bytes)
                dumpResponse(msg)
            }
            else -> {
                error("error: connection timed out")
                exitProcess(1)
            }
        }
    }

    private fun printStatus(status: String) {
        if (!opts.beTerse) error("status: $status")
    }

    private fun dumpResponse(msg: Response) {
        val status = "<recv>"
        if (msg.status?.code == StatusCode.OK) {
            if (opts.beTerse) dumpTerseResponse(msg.media)
            else printMessage(status, msg.media as Message)
        } else printMessage(status, msg)
    }

    private fun dumpTerseResponse(msg: MediaResponse?) {
        info("---\ntitle: ${msg?.title}\nquality:")
        msg?.stream?.forEach {
            info("  profile: ${it.quality?.profile}\n"
                    + "    width: ${it.quality?.width}\n"
                    + "    height: ${it.quality?.height}"
            )
        }
    }

    @OptIn(ExperimentalProtoJson::class)
    private fun printMessage(status: String, msg: Message) {
        val result =
            if (opts.outputJson) msg.encodeToJsonString(jsonConfig)
            else msg.toString()
        printStatus(status)
        info(result)
    }
}

File: lua5/src/demo/jomiel.lua
--
-- -*- coding: utf-8 -*-
--
-- jomiel-examples
--
-- Copyright
--  2019-2021 Toni Gündoğdu
--
--
-- SPDX-License-Identifier: Apache-2.0
--
local Class = require 'middleclass'

local Inspect = require 'inspect'
local DkJSON = require 'dkjson'

local Poller = require 'lzmq.poller'
local ZMQ = require 'lzmq'

local Protoc = require 'protoc'
local PB = require 'pb'

local PkgPath = 'jomiel.protobuf.v1beta1'
Jomiel = Class('Jomiel')

local function loadProtoFiles(options)
  local protoc = Protoc.new()
  protoc.include_imports = true
  protoc:addpath(options.proto_dir)
  protoc:loadfile(PkgPath:gsub('%.', '/') .. '/message.proto')
  --[[
  assert(PB.type 'jomiel.protobuf.v1beta1.Inquiry')
  assert(PB.type 'jomiel.protobuf.v1beta1.MediaInquiry')
  assert(PB.type 'jomiel.protobuf.v1beta1.Response')
  assert(PB.type 'jomiel.protobuf.v1beta1.MediaResponse')
  assert(PB.type 'jomiel.protobuf.v1beta1.StatusCode')
  ]]--
end

function Jomiel:initialize(options)
  loadProtoFiles(options)
  self.sck = ZMQ.context():socket(ZMQ.REQ, { linger = 0 })
  self.opts = options
end

function Jomiel:inquire()
  if next(self.opts.URI) then
    self:_connect()
    for _, uri in pairs(self.opts.URI) do
      self:_sendInquiry(uri)
      self:_receiveResponse()
    end
  else
    io.stderr:write("error: no input URI given\n")
    os.exit(1)
  end
end

function Jomiel:_connect()
  local re = self.opts.router_endpoint
  local to = self.opts.connect_timeout
  self:_printStatus("<connect> %s (timeout=%d)", re, to)
  self.sck:connect(re)
end

function Jomiel:_sendInquiry(uri)
  local msg = { media = { input_uri = uri } }
  if self.opts.be_terse == false then
    self:_printMessage('<send>', msg)
  end
  local bytes = PB.encode(PkgPath .. '.Inquiry', msg)
  self.sck:send(bytes)
end

--[[
local function printSerialized(bytes)
  local len = string.len(bytes)
  local hex = PB.tohex(bytes)
  io.stderr:write(string.format('<recv>: [%d] %s\n', len, hex))
end
]]--

function Jomiel:_receiveResponse()
  local poller = Poller.new(1)

  poller:add(self.sck, ZMQ.POLLIN, function()
    local bytes = self.sck:recv()
    -- printSerialized(bytes)
    local msg = PB.decode(PkgPath .. '.Response', bytes)
    self:_dumpResponse(msg)
  end)

  if poller:poll(self.opts.connect_timeout * 1000) == 0 then
    io.stderr:write('error: connection timed out\n')
    os.exit(1)
  end
end

function Jomiel:_dumpResponse(msg)
  local lookup_path = PkgPath .. '.StatusCode'

  local status_code = PB.enum(lookup_path, msg.status.code)
  local status_ok = PB.enum(lookup_path, 'STATUS_CODE_OK')

  local status = '<recv>'

  if status_code == status_ok then
    if self.opts.be_terse then
      self:_dumpTerseResponse(msg.media)
    else
      self:_printMessage(status, msg.media)
    end
  else
    self:_printMessage(status, msg)
  end
end

function Jomiel:_dumpTerseResponse(msg)
  local function printf(fmt, ...)
    io.write(fmt:format(...) .. '\n')
  end
  printf('---\ntitle: %s\nquality:', msg.title)
  for _, stream in pairs(msg.stream) do
    printf('  profile: %s\n    width: %d\n    height: %d',
      stream.quality.profile,
      stream.quality.width,
      stream.quality.height)
  end
end

function Jomiel:_printStatus(fmt, ...)
  if self.opts.be_terse == false then
    io.stderr:write(fmt:format(...) .. '\n')
  end
end

function Jomiel:_printMessage(status, msg)
  self:_printStatus(status)
  local str
  if self.opts.output_json then
    str = DkJSON.encode(msg, { indent = not self.opts.compact_json })
  else
    str = Inspect(msg)
  end
  print(str)
end

return Jomiel
